src/types.rs

```diff
@@ -72,6 +73,7 @@ pub struct Variable {
 pub struct Expression {
     /// Terms of the expression
     pub(crate) terms: Vec<(r1cs::Variable, Scalar)>,
+    pub(crate) assignment: Option<u64>,
```

Comment:
that's not going to be necessarily a u64, it can be a full-size Scalar.

Reply:
got it :+1: I set it as a `u64` since that's what the range proof function expects, but makes sense to leave as a `Scalar` in the struct!

Reply:
see the `ScalarKind` enum below, this should formalize the requirement that "some scalars must be in range"